File: client/src/render/map-mesh.ts
// Build the static level mesh from MapData. Boxes -> BoxGeometry; ramps ->
// BufferGeometry from the shared solidMeshArrays so render and collider can
// never disagree on a surface.
//
// Look: the bright "clean arena" style (Roblox Rivals) — uniform near-white
// surfaces clad in a tiled panel texture with subtle seams and a grey "+" mark
// at every tile intersection. Edges read from AO + light shadows, not colour.

import * as THREE from 'three';
import { mergeGeometries } from 'three/examples/jsm/utils/BufferGeometryUtils.js';
import type { MapData, Solid, SolidRamp } from '@rivals/shared';
import { solidMeshArrays } from '@rivals/shared';

// Near-white cladding for every surface; the tile texture + shadows carry the
// detail, so a single colour keeps the arena clean like the reference.
const SURFACE_COLOR = 0xdde1e8;
// One tile = this many world units. Sets the grid density across all geometry.
const TILE_WORLD = 2.0;

// Shared tile texture: white panel, faint seam border, and a grey "+" stamped
// across each corner so that — once the tile repeats — every grid intersection
// carries a cross. Built once; each mesh clones it to set its own repeat.
let _tileTex: THREE.Texture | null = null;
function tileTexture(): THREE.Texture {
  if (_tileTex) return _tileTex;
  const px = 256;
  const c = document.createElement('canvas');
  c.width = px;
  c.height = px;
  const ctx = c.getContext('2d')!;

  // Panel base.
  ctx.fillStyle = '#eceef2';
  ctx.fillRect(0, 0, px, px);

  // Seam border (tiles into a continuous thin grid between panels).
  ctx.strokeStyle = 'rgba(150,159,173,0.45)';
  ctx.lineWidth = 2;
  ctx.strokeRect(1, 1, px - 2, px - 2);

  // "+" marks across the four corners → a full cross at every intersection once
  // the texture repeats. Drawn beyond the canvas edge; wrapping reunites them.
  ctx.fillStyle = 'rgba(120,129,146,0.62)';
  const L = 24; // arm half-length
  const T = 7; // arm thickness
  for (const [cx, cy] of [
    [0, 0],
    [px, 0],
    [0, px],
    [px, px],
  ]) {
    ctx.fillRect(cx - L, cy - T / 2, 2 * L, T); // horizontal arm
    ctx.fillRect(cx - T / 2, cy - L, T, 2 * L); // vertical arm
  }

  const tex = new THREE.CanvasTexture(c);
  tex.colorSpace = THREE.SRGBColorSpace;
  tex.wrapS = THREE.RepeatWrapping;
  tex.wrapT = THREE.RepeatWrapping;
  tex.anisotropy = 16; // keep grazing-angle floor tiles crisp instead of washing
  _tileTex = tex;
  return tex;
}

// Multiply a geometry's UVs in place so the tile texture repeats `rx`×`ry` times
// across it. Baking the repeat into the UVs (instead of per-mesh texture.repeat)
// lets every solid share ONE texture + ONE material, so the whole map merges to
// a single draw call. The shared texture stays at wrap-repeat with repeat=(1,1).
function bakeRepeat(geo: THREE.BufferGeometry, rx: number, ry: number): void {
  const uv = geo.getAttribute('uv');
  for (let i = 0; i < uv.count; i++) {
    uv.setXY(i, uv.getX(i) * rx, uv.getY(i) * ry);
  }
  uv.needsUpdate = true;
}

// A solid is treated as "floor" when it is wide/deep and thin in y, sitting low.
function isFloor(s: Solid): boolean {
  const [, sy] = s.size;
  const [sx, , sz] = s.size;
  return s.type === 'box' && sy <= 1.0 && sx >= 8 && sz >= 8;
}

// Returns a world-positioned box geometry with the tile repeat baked into its
// UVs, ready to be merged into the single map mesh.
function buildBox(s: Solid): THREE.BufferGeometry {
  const [sx, sy, sz] = s.size;
  const [px, py, pz] = s.pos;
  const geo = new THREE.BoxGeometry(sx, sy, sz);

  // World-scaled tiling: floors tile across their footprint; uprights tile
  // their widest horizontal span across U and their height up V (BoxGeometry's
  // side faces map U to the wider horizontal axis, V to height).
  const floor = isFloor(s);
  const repeatX = (floor ? sx : Math.max(sx, sz)) / TILE_WORLD;
  const repeatY = (floor ? sz : sy) / TILE_WORLD;
  bakeRepeat(geo, repeatX, repeatY);

  geo.translate(px, py, pz);
  return geo;
}

// solidTriangleIndices winds ramp prisms with handedness that flips by dir (the
// +z / -x prisms wind INWARD), so on FrontSide their faces were culled and the
// ramps rendered see-through. Rather than mask it with DoubleSide, fix it at the
// source: for a convex solid a triangle winds inward iff its geometric normal
// points back toward the solid's centroid — flip those. Then FrontSide draws
// every face and computeVertexNormals yields genuinely-outward normals (correct
// lighting, standard shadows). Zero-alloc, indices rewritten in place. (Kept in
// the render layer to avoid touching the shared collider.)
function fixWindingOutward(geo: THREE.BufferGeometry): void {
  const pos = geo.getAttribute('position');
  const idx = geo.getIndex();
  if (!idx) return;
  let cx = 0;
  let cy = 0;
  let cz = 0;
  for (let i = 0; i < pos.count; i++) {
    cx += pos.getX(i);
    cy += pos.getY(i);
    cz += pos.getZ(i);
  }
  cx /= pos.count;
  cy /= pos.count;
  cz /= pos.count;
  const arr = idx.array as Uint32Array;
  for (let t = 0; t < arr.length; t += 3) {
    const i0 = arr[t];
    const i1 = arr[t + 1];
    const i2 = arr[t + 2];
    const ax = pos.getX(i0);
    const ay = pos.getY(i0);
    const az = pos.getZ(i0);
    const e1x = pos.getX(i1) - ax;
    const e1y = pos.getY(i1) - ay;
    const e1z = pos.getZ(i1) - az;
    const e2x = pos.getX(i2) - ax;
    const e2y = pos.getY(i2) - ay;
    const e2z = pos.getZ(i2) - az;
    // Face normal (e1 × e2) and the triangle-centroid → solid-centroid vector.
    const nx = e1y * e2z - e1z * e2y;
    const ny = e1z * e2x - e1x * e2z;
    const nz = e1x * e2y - e1y * e2x;
    const tcx = (ax + pos.getX(i1) + pos.getX(i2)) / 3 - cx;
    const tcy = (ay + pos.getY(i1) + pos.getY(i2)) / 3 - cy;
    const tcz = (az + pos.getZ(i1) + pos.getZ(i2)) / 3 - cz;
    // Normal pointing inward (toward centroid) → swap to wind outward.
    if (nx * tcx + ny * tcy + nz * tcz < 0) {
      arr[t + 1] = i2;
      arr[t + 2] = i1;
    }
  }
  idx.needsUpdate = true;
}

// Returns a world-space ramp geometry (vertices already world-positioned by the
// shared mesh builder) with planar world-grid UVs, ready to merge.
function buildRamp(s: Solid): THREE.BufferGeometry {
  const { vertices, indices } = solidMeshArrays(s);
  const geo = new THREE.BufferGeometry();
  geo.setAttribute('position', new THREE.BufferAttribute(vertices, 3));
  geo.setIndex(new THREE.BufferAttribute(indices, 1));
  fixWindingOutward(geo); // BEFORE normals — so they come out outward
  geo.computeVertexNormals();

  // Planar UVs from world X/Z so the ramp's tiles line up with the floor grid
  // below it (vertices are already world-space). The slope stretches the tiles
  // slightly along its run — visually continuous and fine for an accent piece.
  // These already carry the world scale, so no repeat baking is needed.
  const pos = geo.getAttribute('position');
  const uv = new Float32Array(pos.count * 2);
  for (let i = 0; i < pos.count; i++) {
    uv[i * 2] = pos.getX(i) / TILE_WORLD;
    uv[i * 2 + 1] = pos.getZ(i) / TILE_WORLD;
  }
  geo.setAttribute('uv', new THREE.BufferAttribute(uv, 2));

  // Each ramp's high (back) wall is coplanar with the block face it climbs, so
  // the two quads z-fought into a shimmering striped band. Sink the RENDER
  // geometry a few cm up-slope (along the ramp's dir axis) so that buried wall
  // slips behind the block face. Cosmetic only — the Rapier collider is
  // untouched, so the sub-cm walk-surface shift is imperceptible.
  const dir = (s as SolidRamp).dir;
  const EPS = 0.04;
  if (dir === '+x') geo.translate(EPS, 0, 0);
  else if (dir === '-x') geo.translate(-EPS, 0, 0);
  else if (dir === '+z') geo.translate(0, 0, EPS);
  else if (dir === '-z') geo.translate(0, 0, -EPS);
  return geo;
}

// Build the whole static level as ONE merged mesh sharing ONE material + ONE
// texture. Every solid's tiling is baked into its UVs (boxes) or carried by
// world-space UVs (ramps), so collapsing them costs nothing visually but turns
// ~18 draw calls + 18 textures + 18 materials into 1 of each. The single mesh
// covers the arena so it's effectively always on-screen — culling it is moot.
export function buildMapMesh(map: MapData): THREE.Group {
  const group = new THREE.Group();
  group.name = 'map';

  const geometries = map.solids.map((s) => (s.type === 'ramp' ? buildRamp(s) : buildBox(s)));
  const merged = mergeGeometries(geometries, false);
  for (const g of geometries) g.dispose();

  const material = new THREE.MeshStandardMaterial({
    color: SURFACE_COLOR,
    roughness: 0.88,
    metalness: 0.0,
    map: tileTexture(), // shared singleton; repeat stays (1,1), UVs carry scale
  });

  const mesh = new THREE.Mesh(merged, material);
  mesh.name = 'map-merged';
  group.add(mesh);
  return group;
}

File: godot/src/Client/Render/MapMesh.cs
// Build the static level mesh from MapData. Boxes -> BoxMesh; ramps ->
// ArrayMesh from the shared solidMeshArrays so render and collider never disagree
// about a surface. The bright "clean arena" look (Roblox Rivals): near-white
// surfaces with a tiled panel texture (triplanar), depth from AO + light shadows.

using Godot;
using WebRivals.Shared;

namespace WebRivals.Client.Render;

public static class MapMesh
{
    private static readonly Color SurfaceColor = new(0.867f, 0.882f, 0.910f);
    private const float TileWorld = 2.0f;

    public static Node3D Build(MapData map)
    {
        var root = new Node3D { Name = "Map" };
        var tex = TileTexture();

        var mat = new StandardMaterial3D
        {
            AlbedoColor = SurfaceColor,
            Roughness = 0.88f,
            Metallic = 0.0f,
            AlbedoTexture = tex,
            Uv1Triplanar = true,
            Uv1Scale = new Vector3(1f / TileWorld, 1f / TileWorld, 1f / TileWorld),
            TextureFilter = BaseMaterial3D.TextureFilterEnum.LinearWithMipmapsAnisotropic,
        };

        // Ramps render double-sided so winding differences never read as see-through.
        var rampMat = (StandardMaterial3D)mat.Duplicate();
        rampMat.CullMode = BaseMaterial3D.CullModeEnum.Disabled;

        foreach (var s in map.Solids)
        {
            MeshInstance3D mi;
            if (s.Type == SolidType.Box)
            {
                var box = new BoxMesh { Size = new Vector3((float)s.Size[0], (float)s.Size[1], (float)s.Size[2]) };
                mi = new MeshInstance3D { Mesh = box, MaterialOverride = mat };
                mi.Position = new Vector3((float)s.Pos[0], (float)s.Pos[1], (float)s.Pos[2]);
            }
            else
            {
                mi = new MeshInstance3D { Mesh = BuildRampMesh(s), MaterialOverride = rampMat };
                // Ramp vertices are already world-space.
            }
            mi.CastShadow = GeometryInstance3D.ShadowCastingSetting.On;
            root.AddChild(mi);
        }

        return root;
    }

    private static ArrayMesh BuildRampMesh(Solid s)
    {
        var (verts, indices) = Geometry.SolidMeshArrays(s);
        var st = new SurfaceTool();
        st.Begin(Mesh.PrimitiveType.Triangles);
        for (int i = 0; i < indices.Length; i++)
        {
            int vi = indices[i] * 3;
            st.SetUV(new Vector2(verts[vi] / TileWorld, verts[vi + 2] / TileWorld));
            st.AddVertex(new Vector3(verts[vi], verts[vi + 1], verts[vi + 2]));
        }
        st.GenerateNormals();
        st.GenerateTangents();
        return st.Commit();
    }

    // White panel with a faint seam border and a grey "+" at each corner so the
    // repeated texture stamps a cross at every tile intersection.
    private static ImageTexture TileTexture()
    {
        const int px = 256;
        var img = Image.CreateEmpty(px, px, false, Image.Format.Rgba8);
        img.Fill(new Color(0.925f, 0.933f, 0.949f));

        var seam = new Color(0.588f, 0.624f, 0.678f, 0.45f);
        // Seam border (1px frame).
        for (int i = 0; i < px; i++)
        {
            img.SetPixel(i, 0, seam); img.SetPixel(i, 1, seam);
            img.SetPixel(i, px - 1, seam); img.SetPixel(i, px - 2, seam);
            img.SetPixel(0, i, seam); img.SetPixel(1, i, seam);
            img.SetPixel(px - 1, i, seam); img.SetPixel(px - 2, i, seam);
        }
        // Corner "+" arms (so neighbours form a full cross when tiled).
        var cross = new Color(0.471f, 0.506f, 0.573f, 0.62f);
        const int L = 24, T = 7;
        (int, int)[] corners = { (0, 0), (px, 0), (0, px), (px, px) };
        foreach (var (cx, cy) in corners)
        {
            for (int dx = -L; dx < L; dx++)
                for (int dy = -T / 2; dy <= T / 2; dy++)
                    Plot(img, cx + dx, cy + dy, cross);
            for (int dy = -L; dy < L; dy++)
                for (int dx = -T / 2; dx <= T / 2; dx++)
                    Plot(img, cx + dx, cy + dy, cross);
        }

        img.GenerateMipmaps();
        return ImageTexture.CreateFromImage(img);
    }

    private static void Plot(Image img, int x, int y, Color c)
    {
        if (x < 0 || y < 0 || x >= img.GetWidth() || y >= img.GetHeight()) return;
        img.SetPixel(x, y, c);
    }
}
